libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c

```diff
@@ -354,790 +354,826 @@ void prvFindObjectInListByLabel( uint8_t * pcLabel,
+
+                if( xCertificateType != CKC_X_509 )
+                {
+                    xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+                }
+
+                break;
+
+            case ( CKA_TOKEN ):
+                memcpy( &xBool, xAttribute.pValue, sizeof( CK_BBOOL ) );
+
+                if( xBool != CK_TRUE )
+                {
+                    PKCS11_PRINT( ( "ERROR: Only token key object is supported. \r\n" ) );
+                    xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+                }
+
+                break;
+
+            case ( CKA_CLASS ):
+                pxClass = &pxTemplate[ ulIndex ];
+                break;
+
+            case ( CKA_SUBJECT ):
+
+                /* Do nothing.  This was already parsed out of the template previously. */
+                break;
+
+            default:
+                xResult = CKR_TEMPLATE_INCONSISTENT;
+                break;
+        }
+    }
+
+    if( ( pxCertificateValue == NULL ) || ( pxLabel == NULL ) )
+    {
+        xResult = CKR_TEMPLATE_INCOMPLETE;
+    }
+
+    if( xResult == CKR_OK )
+    {
+        xPalHandle = PKCS11_PSA_SaveObject( pxClass, pxLabel, pxCertificateValue, xCertificateLength, NULL );
+
+        if( xPalHandle == 0 ) /*Invalid handle. */
+        {
+            xResult = CKR_DEVICE_MEMORY;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        xResult = prvAddObjectToList( xPalHandle, pxObject, pxLabel->pValue, pxLabel->ulValueLen );
+        /* TODO: If this fails, should the object be wiped back out of flash?  But what if that fails?!?!? */
+    }
+
+    return xResult;
 }
 
-#define PKCS11_INVALID_KEY_TYPE	( ( CK_KEY_TYPE ) 0xFFFFFFFF )
+#define PKCS11_INVALID_KEY_TYPE    ( ( CK_KEY_TYPE ) 0xFFFFFFFF )
 
 /* Helper to search an attribute for the key type attribute. */
 void prvGetKeyType( CK_KEY_TYPE * pxKeyType,
-					CK_ATTRIBUTE_PTR pxTemplate,
-					CK_ULONG ulCount )
+                    CK_ATTRIBUTE_PTR pxTemplate,
+                    CK_ULONG ulCount )
 {
-	uint32_t ulIndex;
-	CK_ATTRIBUTE xAttribute;
+    uint32_t ulIndex;
+    CK_ATTRIBUTE xAttribute;
 
-	*pxKeyType = PKCS11_INVALID_KEY_TYPE;
+    *pxKeyType = PKCS11_INVALID_KEY_TYPE;
 
-	for( ulIndex = 0; ulIndex < ulCount; ulIndex++ )
-	{
-		xAttribute = pxTemplate[ ulIndex ];
+    for( ulIndex = 0; ulIndex < ulCount; ulIndex++ )
+    {
+        xAttribute = pxTemplate[ ulIndex ];
 
-		if( xAttribute.type == CKA_KEY_TYPE )
-		{
-			memcpy( pxKeyType, xAttribute.pValue, sizeof( CK_KEY_TYPE ) );
-			break;
-		}
-	}
+        if( xAttribute.type == CKA_KEY_TYPE )
+        {
+            memcpy( pxKeyType, xAttribute.pValue, sizeof( CK_KEY_TYPE ) );
+            break;
+        }
+    }
 }
 
 /* Helper to search a template for the label attribute. */
 void prvGetLabel( CK_ATTRIBUTE_PTR * ppxLabel,
-				  CK_ATTRIBUTE_PTR pxTemplate,
-				  CK_ULONG ulCount )
+                  CK_ATTRIBUTE_PTR pxTemplate,
```

Comment:
The types of the first input parameter of the two functions are not the same. It would make the parameter unfriendly to users if combine the two functions together. And the code size is not that much. So I prefer to keep it as it is.